### PR TITLE
Modernize cmake; make cmake compatible with git submoduling

### PR DESCRIPTION
**Pull request by jlblancoc**
This PR does:
* Use `GNUInstallDirs` to correctly install files in their standard directories (multiarch-compatible).
* Add `install(EXPORT ...` to cmake scripts to allow better use of exported targets.
* Add build options to allow not building parts of the projects (defaults=as it was before)
* Automatic detection of `libnabo` if built as part of the same source tree (e.g. git submodules).

### CODE REVIEW

**Comment by pomerlef** [comment]
ok to test

File: CMakeLists.txt [hunk 1]
```diff
@@ -29,6 +29,10 @@ else (CMAKE_VERSION VERSION_LESS 3.0.0)
 	project(libpointmatcher VERSION ${POINTMATCHER_PROJECT_VERSION})
 endif (CMAKE_VERSION VERSION_LESS 3.0.0)
 
+set(LIB_NAME pointmatcher) # same than PROJECT_NAME, w/o the "lib" prefix
```
Comment (YoshuaNava) [inline CMakeLists.txt:32]: Two questions:
* Why are you defining this alternate library name?
* If it's necessary, why not re-using PROJECT_NAME?

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,6 +29,10 @@ else (CMAKE_VERSION VERSION_LESS 3.0.0)
...
+
+include(GNUInstallDirs) # populate CMAKE_INSTALL_{LIB,BIN}DIR
```
Comment (YoshuaNava) [inline CMakeLists.txt:34]: Can we move all the includes to a single block? Or at least most of them.

Also, would it make sense to use include_directories?

File: CMakeLists.txt [hunk 3]
```diff
@@ -215,9 +227,8 @@ else(USE_SYSTEM_YAML_CPP)
         #get_property(yaml-cpp-pm_INCLUDE TARGET yaml-cpp-pm PROPERTY INCLUDE_DIRECTORIES)
         #include_directories(${yaml-cpp-pm_INCLUDE})
 
-        get_property(yaml-cpp-pm_LIB TARGET yaml-cpp-pm PROPERTY LOCATION)
-        set (EXTERNAL_LIBS ${EXTERNAL_LIBS} ${yaml-cpp-pm_LIB} )
-        set (EXTRA_DEPS ${EXTRA_DEPS} yaml-cpp-pm)
+        list(APPEND EXTERNAL_LIBS $<TARGET_FILE:yaml-cpp-pm>)
```
Comment (YoshuaNava) [inline CMakeLists.txt:230]: This generator only works during build time. Does that fulfill the purpose of the expression?

If I'm not mistaken, before we were fetching the location of yaml_cpp. Is it possible to replace that with a standard cmake macro as find_package?

File: CMakeLists.txt [hunk 4]
```diff
@@ -343,37 +354,31 @@ set(POINTMATCHER_SRC
...
-		install(TARGETS pointmatcher ARCHIVE DESTINATION ${INSTALL_LIB_DIR})
-	endif(SHARED_LIBS)
-endif(APPLE OR WIN32)
-
-target_include_directories(pointmatcher PRIVATE ${CMAKE_SOURCE_DIR}/pointmatcher)
-target_include_directories(pointmatcher PRIVATE ${CMAKE_SOURCE_DIR}/pointmatcher/DataPointsFilters)
-target_include_directories(pointmatcher PRIVATE ${CMAKE_SOURCE_DIR}/pointmatcher/DataPointsFilters/utils)
+
+# In CMake >=3.4 we can easily build shared libraries in Mac and Windows.
+# Don't make any distinction between OSes:
```
Comment (YoshuaNava) [inline CMakeLists.txt:362]: Cool! 🥳

Comment (YoshuaNava) [reply CMakeLists.txt:362]: I would rephrase this comment as
'No need to distinguish between operating systems while building targets'

File: CMakeLists.txt [hunk 5]
```diff
@@ -343,37 +354,31 @@ set(POINTMATCHER_SRC
...
+target_include_directories(pointmatcher PUBLIC
+  $<INSTALL_INTERFACE:>
+  $<INSTALL_INTERFACE:pointmatcher>
+  $<INSTALL_INTERFACE:pointmatcher/DataPointsFilters>
+  $<INSTALL_INTERFACE:pointmatcher/DataPointsFilters/utils>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/pointmatcher>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/pointmatcher/DataPointsFilters>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/pointmatcher/DataPointsFilters/utils>
+)
```
Comment (YoshuaNava) [inline CMakeLists.txt:381]: Is it necessary to specify all the children dirs for datapointsfilters?

File: CMakeLists.txt [hunk 6]
```diff
@@ -343,37 +354,31 @@ set(POINTMATCHER_SRC
...
-target_include_directories(pointmatcher PRIVATE ${CMAKE_SOURCE_DIR}/pointmatcher)
-target_include_directories(pointmatcher PRIVATE ${CMAKE_SOURCE_DIR}/pointmatcher/DataPointsFilters)
-target_include_directories(pointmatcher PRIVATE ${CMAKE_SOURCE_DIR}/pointmatcher/DataPointsFilters/utils)
+
+# In CMake >=3.4 we can easily build shared libraries in Mac and Windows.
+# Don't make any distinction between OSes:
+
+add_library(${LIB_NAME} ${POINTMATCHER_SRC} ${POINTMATCHER_HEADERS} )
+
+install(TARGETS ${LIB_NAME} EXPORT ${PROJECT_NAME}-config
```
Comment (YoshuaNava) [inline CMakeLists.txt:366]: Given that people don't normally install when they build, can we move the install clause below the target_include_directories macro?

File: CMakeLists.txt [hunk 7]
```diff
@@ -454,16 +466,21 @@ endif()
...
+install(EXPORT ${PROJECT_NAME}-config DESTINATION share/${PROJECT_NAME}/cmake)
 
-# Register the local build in case one doesn't use "make install"
-export(PACKAGE libpointmatcher)
+# make project importable from build_dir:
+export(
+	TARGETS ${LIB_NAME}
+	FILE ${CMAKE_BINARY_DIR}/${PROJECT_NAME}-config.cmake
+)
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${LIB_NAME})
```
Comment (YoshuaNava) [inline CMakeLists.txt:477]: Would it be possible to instead reuse PROJECT_NAME?

File: CMakeLists.txt (hunk 6 above)
Comment (jlblancoc) [reply CMakeLists.txt:366]: > Given that people don't normally install when they build,

You would be surprised to see how many people still do it! ;-)
Thanks for the comments on improvements, I will go over them asap.

Comment (YoshuaNava) [reply CMakeLists.txt:366]: > You would be surprised to see how many people still do it! ;-)

Sorry, that's true. :slightly_smiling_face: 

I'm more in favor of keeping a cmake workflow in which you first define/find depends -> global includes -> target definition with local includes -> install -> other actions (test / lint / etc). This might be just a personal preference, so feel free to disregard it.

File: CMakeLists.txt (hunk 1 above)
Comment (jlblancoc) [reply CMakeLists.txt:34]: It's named for similarity with the related project libnabo: 
https://github.com/ethz-asl/libnabo/blob/master/CMakeLists.txt#L7-L8

The project is named libpointmatcher, the targets were defined like `add_library(pointmatcher)` (!). 
So, to avoid the repeated use of the two names, I preferred to define that variable. 
For sure, using just PROJECT_NAME would be better, but that would probably require many changes.

File: CMakeLists.txt (hunk 2 above)
Comment (jlblancoc) [reply CMakeLists.txt:36]: Done!

File: CMakeLists.txt (hunk 3 above)
Comment (jlblancoc) [reply CMakeLists.txt:232]: yes, this is another obscure thing that would require much more work to make better: PM is using an embedded version of the yamp-cpp library, that's the reason of this expression. 
Look at "USE_SYSTEM_YAML_CPP"...

There are *many* things that could be done in more modern-cmake ways, but I don't have the cycles to work on that now, and it has the risk of breaking existing stuff :-)

File: CMakeLists.txt (hunk 4 above)
Comment (jlblancoc) [reply CMakeLists.txt:364]: Done!

File: CMakeLists.txt (hunk 6 above)
Comment (jlblancoc) [reply CMakeLists.txt:368]: > I'm more in favor of keeping a cmake workflow in which you first define/find depends -> global includes -> target definition with local includes -> install -> other actions (test / lint / etc). This might be just a personal preference, so feel free to disregard it.

I am, too. I tried in my upcoming commit to also reorder it a little bit closer to that ideal.

File: CMakeLists.txt (hunk 7 above)
Comment (jlblancoc) [reply CMakeLists.txt:479]: I guess you refer to the LIB_NAME? Notice my comment above: one has the "lib" prefix, the other doesn't.

**Comment by jlblancoc** [comment]
> Have you also tested that the library builds with ROS? In merged and linked devel (--no-install and --install flags when using catkin build)

I tried with catkin, and it failed due to failed to find "nabo/nabo.h"... But I'm not sure it's due to these changes. Perhaps it's assumed that the user should do a "sudo make install" after building "libnabo"? I will re-check this area, anyway, another day.

**Comment by YoshuaNava** [comment]
Hi @jlblancoc,
Thank you for addressing all the comments.

For the ones that require improvements that go beyond the scope of this PR, could you open issues? So we can actively keep track of them.

Thank you, and have a nice weekend :)

Yoshua

**Comment by jlblancoc** [comment]
All done: it was tested to build both, as a standalone project, and with "catkin build". 

Note that there's another related PR in the queue: https://github.com/ethz-asl/libnabo/pull/103

File: CMakeLists.txt (hunk 5 above)
Comment (jlblancoc) [reply CMakeLists.txt:356]: It was like that originally. 
But I didn't like it neither. Fixed in: https://github.com/MOLAorg/libpointmatcher/commit/83107909dfe3bc02c38d9f8d556fc5d4b2d8954e

**Comment by YoshuaNava** [comment]
Nice! Thank you.

**Comment by jlblancoc** [comment]
I'm alive ;-)
Will come back soon, sorry...

**Comment by jlblancoc** [comment]
Please, test and merge https://github.com/ethz-asl/libnabo/pull/103 first so this PR makes sure of using the updated version of libnabo too.

**Comment by pomerlef** [comment]
@jlblancoc so I guess we can continue on this PR. Could you update your branch?

**Comment by jlblancoc** [comment]
@pomerlef After the last commit, it builds cleanly locally... let's see what happens with Jenkins.

**Comment by pomerlef** [comment]
ok to test

**Comment by pomerlef** [comment]
the branch is still out-of-date. Here is the error log:

```
/home/jenkins/workspace/pointmatcher@2/label/ubuntu-xenial/pointmatcher/MatchersImpl.h:41:23: fatal error: nabo/nabo.h: No such file or directory
compilation terminated.
[ 31%] Building CXX object CMakeFiles/pointmatcher.dir/pointmatcher/Registrar.cpp.o
CMakeFiles/pointmatcher.dir/build.make:158: recipe for target 'CMakeFiles/pointmatcher.dir/pointmatcher/ICP.cpp.o' failed

make[2]: *** [CMakeFiles/pointmatcher.dir/pointmatcher/ICP.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/jenkins/workspace/pointmatcher@2/label/ubuntu-xenial/pointmatcher/Registry.cpp:41:0:

/home/jenkins/workspace/pointmatcher@2/label/ubuntu-xenial/pointmatcher/MatchersImpl.h:41:23: fatal error: nabo/nabo.h: No such file or directory
compilation terminated.
CMakeFiles/pointmatcher.dir/build.make:182: recipe for target 'CMakeFiles/pointmatcher.dir/pointmatcher/Registry.cpp.o' failed
```

**Comment by jlblancoc** [comment]
I just integrated @peci1 's fixes (thanks!) into this PR via a cherry-pick. 

The order of operations to get this sorted out is: 
1) Merge https://github.com/ethz-asl/libnabo/pull/106
2) Then trigger another build of this PR to verify it builds in Jenkins. 
3) Merge this PR
4) #404 can be closed since its changes are already applied on top of this PR.

**Comment by pomerlef** [comment]
ok to test

**Comment by pomerlef** [comment]
@jlblancoc could you bring your branch to master so I can merge?

**Comment by jlblancoc** [comment]
I will

**Comment by jlblancoc** [comment]
rebase done.
Let's wait for Jenkins...

**Comment by iabaldwin** [comment]
Builds nicely again - thanks @jlblancoc @pomerlef

**Comment by peci1** [comment]
Not yet there...

```
$ cat devel/lib/pkgconfig/libpointmatcher.pc 
prefix=/media/data/subt/mapper/devel
exec_prefix=/media/data/subt/mapper/devel
libdir=${prefix}/lib
includedir=${prefix}/include

Name: libpointmatcher
Description: See https://github.com/ethz-asl/libpointmatcher
Version: 1.3.1
Libs: -L${libdir}  /media/data/subt/mapper/devel/lib/$<TARGET_FILE:libpointmatcher> -l$<TARGET_FILE:yaml-cpp-pm>
Cflags: -I${includedir}/
```

```
$ cat devel/share/libpointmatcher/cmake/libpointmatcherConfig.cmake 
# - Config file for the libpointmatcher package
# It defines the following variables
#  libpointmatcher_INCLUDE_DIRS - include directories for pointmatcher
#  libpointmatcher_LIBRARIES    - libraries to link against

include(CMakeFindDependencyMacro)
find_dependency(libnabo REQUIRED)

# Compute paths
get_filename_component(POINTMATCHER_CMAKE_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
set(libpointmatcher_INCLUDE_DIRS "/media/data/subt/mapper/devel/include;/usr/include/eigen3")
set(LIBPOINTMATCHER_INCLUDE_DIRS "/media/data/subt/mapper/devel/include;/usr/include/eigen3")
set(pointmatcher_INCLUDE_DIRS "/media/data/subt/mapper/devel/include;/usr/include/eigen3")
set(POINTMATCHER_INCLUDE_DIRS "/media/data/subt/mapper/devel/include;/usr/include/eigen3")

set(libpointmatcher_LIBRARIES "/media/data/subt/mapper/devel/lib/$<TARGET_FILE:libpointmatcher>;$<TARGET_FILE:yaml-cpp-pm>")
set(LIBPOINTMATCHER_LIBRARIES "/media/data/subt/mapper/devel/lib/$<TARGET_FILE:libpointmatcher>;$<TARGET_FILE:yaml-cpp-pm>")
set(pointmatcher_LIBRARIES "/media/data/subt/mapper/devel/lib/$<TARGET_FILE:libpointmatcher>;$<TARGET_FILE:yaml-cpp-pm>")
set(POINTMATCHER_LIBRARIES "/media/data/subt/mapper/devel/lib/$<TARGET_FILE:libpointmatcher>;$<TARGET_FILE:yaml-cpp-pm>")

# This causes catkin simple to link against these libraries
set(libpointmatcher_FOUND_CATKIN_PROJECT true)
set(LIBPOINTMATCHER_FOUND_CATKIN_PROJECT true)
set(pointmatcher_FOUND_CATKIN_PROJECT true)
set(POINTMATCHER_FOUND_CATKIN_PROJECT true)
```

```
Errors     << libpointmatcher_ros:check /media/data/subt/mapper/logs/libpointmatcher_ros/build.check.038.log                                                                
/media/data/subt/mapper/devel/lib/$<TARGET_FILE:libpointmatcher>;$<TARGET_FILE:yaml-cpp-pm>
CMake Error at /media/data/subt/mapper/src/ethzasl_icp_mapping/libpointmatcher_ros/CMakeLists.txt:62 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_FILE:libpointmatcher>

  No target "libpointmatcher"


CMake Error at /media/data/subt/mapper/src/ethzasl_icp_mapping/libpointmatcher_ros/CMakeLists.txt:62 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_FILE:yaml-cpp-pm>

  No target "yaml-cpp-pm"


make: *** [cmake_check_build_system] Error 1
```

We need to find a way for the generator expressions to actually get evaluated.

**Comment by peci1** [comment]
See #407.
